File: frontend/src/app/services/permission.service.ts
import {inject, Injectable, signal} from '@angular/core';
import { HttpService } from './http.service';
import { PermissionSetDto } from '../dtos/PermissionSetDto';
import {Observable, of, tap, throwError} from 'rxjs';
import {catchError} from 'rxjs/operators';
import {LoadingService} from "./loading.service";

@Injectable({
  providedIn: 'root'
})
export class PermissionService {
  private permissions = signal<PermissionSetDto | null>(null);
  private loading = false;
  private http = inject(HttpService);
  private loadingService = inject(LoadingService);

  fetchPermissions(force = false, informLoadingService: boolean = true): Observable<PermissionSetDto> {
    if (this.loading && !force) {
      return of(this.permissions() || {
        admin: false,
        environments: 'none',
        environment: { admin: false, all: 'none', specific: {} },
        clients: 'none',
        settings: 'none',
        features: 'none',
        resolved: false
      });
    }

    if (informLoadingService) {
      this.loadingService.setApplicationLoading(true);
    }

    this.loading = true;
    return this.http.get<PermissionSetDto>('/api/auth/permissions').pipe(
      tap(perms => {
        this.permissions.set(perms);
        this.loading = false;
        if (informLoadingService) {
          this.loadingService.setApplicationLoading(false);
        }
      }),
      catchError(err => {
        this.loading = false;
        if (informLoadingService) {
          this.loadingService.setApplicationLoading(false);
        }
        return throwError(() => err);
      })
    );
  }

  refreshPermissions(): Observable<PermissionSetDto> {
    return this.fetchPermissions(true);
  }

  clearPermissions() {
    this.permissions.set(null);
  }

  hasPermission(resource: string, action: 'read' | 'write'): boolean {
    const perms = this.permissions();
    if (!perms) return false;
    if (perms.admin) return true;

    let level: 'none' | 'read' | 'write' = 'none';
    switch (resource) {
      case 'environments': level = perms.environments; break;
      case 'clients': level = perms.clients; break;
      case 'settings': level = perms.settings; break;
      case 'features': level = perms.features; break;
    }

    return this.checkAccess(level, action);
  }

  hasEnvPermission(envName: string, action: 'read' | 'write'): boolean {
    const perms = this.permissions();
    if (!perms) return false;
    if (perms.admin || perms.environment.admin) return true;

    if (this.checkAccess(perms.environment.all, action)) return true;

    const specificLevel = perms.environment.specific[envName] || 'none';
    return this.checkAccess(specificLevel, action);
  }

  private checkAccess(level: 'none' | 'read' | 'write', action: 'read' | 'write'): boolean {
    if (action === 'write') return level === 'write';
    if (action === 'read') return level === 'read' || level === 'write';
    return false;
  }
}

File: backend/backend/src/main/kotlin/com/vivid/backend/service/PermissionService.kt
package com.vivid.backend.service

import com.vivid.backend.api.web.dto.EnvironmentPermissionsDto
import com.vivid.backend.api.web.dto.PermissionSetDto
import com.vivid.backend.domain.entity.EnvironmentEntity
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.oauth2.jwt.Jwt
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken
import org.springframework.stereotype.Service

@Service
class PermissionService(
    private val properties: PermissionProperties,
) {

    private val adminRole = properties.rolePrefix + properties.adminRole
    private val superUserPermissions = PermissionSetDto(
        admin = true,
        environments = "write",
        environment = EnvironmentPermissionsDto(admin = true, all = "write")
    )

    fun filterVisibleEnvironments(environments: List<EnvironmentEntity>): List<EnvironmentEntity> {
        val perms = getEffectivePermissions()
        if (perms.admin || perms.environment.admin) return environments

        val allLevel = perms.environment.all
        if (allLevel == "read" || allLevel == "write") return environments

        return environments.filter { env ->
            val specificLevel = perms.environment.specific[env.name] ?: "none"
            specificLevel == "read" || specificLevel == "write"
        }
    }

    fun getEffectivePermissions(): PermissionSetDto {
        val authentication = SecurityContextHolder.getContext().authentication
        if (authentication !is JwtAuthenticationToken) {
            return PermissionSetDto(resolved = false)
        }

        val jwt = authentication.token
        val roles = extractRoles(jwt)

        if (properties.rootRole != null && roles.contains(properties.rootRole)) {
            return superUserPermissions
        }

        if (roles.contains(adminRole)) {
            return superUserPermissions
        }

        val envAdmin = roles.contains("${properties.rolePrefix}env:admin")
        val envAllWrite = envAdmin || roles.contains("${properties.rolePrefix}env:all:write")
        val envAllRead = envAllWrite || roles.contains("${properties.rolePrefix}env:all:read")

        val specificEnv = mutableMapOf<String, String>()
        roles.forEach { role ->
            if (role.startsWith("${properties.rolePrefix}env:")) {
                val parts = role.split(":")
                if (parts.size == 3) {
                    val envId = parts[1]
                    val action = parts[2] // "read" or "write"
                    if (envId != "all" && envId != "admin") {
                        val current = specificEnv[envId]
                        if (current != "write") {
                            specificEnv[envId] = action
                        }
                    }
                }
            }
        }

        return PermissionSetDto(
            admin = false,
            environments = getAccessLevel(roles, "environments"),
            environment = EnvironmentPermissionsDto(
                admin = envAdmin,
                all = if (envAllWrite) "write" else if (envAllRead) "read" else "none",
                specific = specificEnv
            )
        )
    }

    private fun getAccessLevel(roles: Collection<String>, resource: String): String {
        return when {
            roles.contains("${properties.rolePrefix}$resource:write") -> "write"
            roles.contains("$resource:read") -> "read"
            else -> properties.defaultVisibility[resource] ?: "none"
        }
    }

    private fun extractRoles(jwt: Jwt): Set<String> {
        val realmAccess = jwt.getClaim<Map<String, Any>>("realm_access")
        val roles = realmAccess?.get("roles") as? List<String> ?: emptyList()
        return roles.toSet()
    }

    fun hasPermission(resource: String, action: String): Boolean {
        val perms = getEffectivePermissions()
        if (perms.admin) return true

        val level = when (resource) {
            "environments" -> perms.environments
            else -> "none"
        }

        return checkAccess(level, action)
    }

    fun hasEnvPermission(envId: String, action: String): Boolean {
        val perms = getEffectivePermissions()
        if (perms.admin || perms.environment.admin) return true

        val allLevel = perms.environment.all
        if (checkAccess(allLevel, action)) return true

        val specificLevel = perms.environment.specific[envId] ?: "none"
        return checkAccess(specificLevel, action)
    }

    private fun checkAccess(level: String, action: String): Boolean {
        return when (action) {
            "write" -> level == "write"
            "read" -> level == "write" || level == "read"
            else -> false
        }
    }
}
